packages/core/src/external/mpi/match-patients.ts

```diff
@@ -0,0 +1,171 @@
+  patient: PatientDataMPI
+): boolean => {
+  if (
+    demo.personalIdentifiers &&
+    demo.personalIdentifiers.length > 0 &&
+    intersectionWith(patient.personalIdentifiers, demo.personalIdentifiers, isEqual).length > 0
+  ) {
+    return true;
+  }
+  return false;
+};
+
+/**
+ * This function checks if the patient has any contact details (phone or email) that match the demo. The idea of this rule
+ * us that contact info is a very strong unique identifier if they do match, but doesn't necessarily mean that the patients
+ * don't match if they don't match.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any contact details that match the demo.
+ */
+export const matchingContactDetailsRule = (
+  demo: PatientDataMPI,
+  patient: PatientDataMPI
+): boolean => {
+  // Check for matching phone numbers
+  if (
+    demo.contact &&
+    demo.contact.length > 0 &&
+    intersectionWith(patient.contact, demo.contact, (a, b) => a.phone === b.phone).length > 0
+  ) {
+    return true;
+  }
+  // Check for matching emails
+  if (
+    demo.contact &&
+    demo.contact.length > 0 &&
+    intersectionWith(patient.contact, demo.contact, (a, b) => a.email === b.email).length > 0
+  ) {
+    return true;
+  }
+  return false;
+};
+
+/**
+ * This function calculates the similarity between two patients using the Jaro-Winkler algorithm.
+ * It returns a score between 0 and 1, where 1 means the patients are identical. We calculate scores
+ * for the following fields: First Name, Last Name, Address Line 1, Address Line 2, City, State, Country, Zipcode.
+ * This function won't be called if gender and DOB are not identical, so that is a given.
+ * @param patient1
+ * @param patient2
+ * @returns The average of the similarity scores for each field.
+ */
+export const jaroWinklerSimilarity = (
+  patient1: PatientDataMPI,
+  patient2: PatientDataMPI,
+  threshold: number
+): boolean => {
+  let score = 0;
+  let fieldCount = 0;
+  const similarityScores: { [key: string]: [number, string?, string?] } = {};
+
+  const addScore = (field: string, value1: string, value2: string) => {
+    const similarity = jaroWinkler(value1, value2);
+    similarityScores[field] = [similarity, value1, value2];
+    score += similarity;
+    fieldCount += 1;
+  };
+
+  addScore("First Name", patient1.firstName, patient2.firstName);
+  addScore("Last Name", patient1.lastName, patient2.lastName);
+
+  // Calculate similarity for addresses
+  const address1 = patient1.address && patient1.address.length > 0 ? patient1.address[0] : null;
+  const address2 = patient2.address && patient2.address.length > 0 ? patient2.address[0] : null;
+
+  if (address1 && address2) {
+    if (address1.addressLine1 && address2.addressLine1) {
+      addScore("Address Line 1", address1.addressLine1, address2.addressLine1);
+    }
+
+    if (address1.addressLine2 && address2.addressLine2) {
+      addScore("Address Line 2", address1.addressLine2, address2.addressLine2);
+    }
```

Comment:
This is not the actual desired behavior here when we have multiple addresses. Permutations comparisons would be better. This is one of many instances of the general principle addressed in go live ticket: https://github.com/metriport/metriport/issues/1358